src/libxtp/gwbse/gw.cc

```diff
@@ -152,35 +163,255 @@ bool GW::Converged(const Eigen::VectorXd& e1, const Eigen::VectorXd& e2,
+    const int nx = _opt.gw_sc_root_finder_steps;      // Steps
+    const double inf = std::numeric_limits<double>::infinity();
+    // Grid refinement
+    const double f_refine = _opt.gw_sc_root_finder_refine;
+    const bool b_refine = f_refine > 0.0 && f_refine < 1.0;
+    const double dx_min = _opt.g_sc_limit * (nx - 1.0) / (2.0 * rx0);
+    const int n_refine_max = std::ceil(std::log(dx_min) / std::log(f_refine));
+
+    int n_refine = 0;  // Refinement number
+    for (Index i_freq = 0; i_freq < _opt.g_sc_max_iterations; ++i_freq) {
+      // Grid refinement
+      const double rx =
+          rx0 * std::pow(f_refine, n_refine);  // std::min(n_refine_max,
+                                               // n_refine));
+      const double dx = (2.0 * rx) / (nx - 1.0);
+      // Prepare vectors
+      const Eigen::VectorXd xx_off = Eigen::VectorXd::LinSpaced(nx, -rx, +rx);
+      const Eigen::VectorXd sx_vxc = _Sigma_x.diagonal() - _vxc.diagonal();
+      // Evaluate sigma_c on all grid points
+      Eigen::MatrixXd xx =
+          Eigen::MatrixXd::Zero(nx, _qptotal);  // TODO: Do not cache this
+      Eigen::MatrixXd fx = Eigen::MatrixXd::Zero(nx, _qptotal);
+      if (tools::globals::verbose) {
+        XTP_LOG_SAVE(logDEBUG, _log)
+            << TimeStamp() << " Grid:"
+            << " nx: " << nx << " rx: " << rx << " dx: " << dx << std::flush;
+      }
+      // TODO: Is it faster to first fill the columns and then transpose the
+      // entire matrix?
+      for (Index i_x = 0; i_x < nx; ++i_x) {
+        Eigen::VectorXd xx_cur = frequencies.array() + xx_off[i_x];
+        xx.row(i_x) = xx_cur;
+        fx.row(i_x) = _sigma->CalcCorrelationDiag(xx_cur);
+      }  // Grid point i_x
+      // For each state, find best root
+      Eigen::VectorXd root_values = Eigen::VectorXd::Zero(_qptotal);
+      Eigen::VectorXd root_scores = Eigen::VectorXd::Zero(_qptotal);
+      // TODO: Multi-thread?
+      for (Index i_qp = 0; i_qp < _qptotal; ++i_qp) {
+        //    e_GW = sigma_c(e_GW) + sigma_x - v_xc + e_DFT
+        // =>    0 = sigma_c(e_GW) + sigma_x - v_xc + e_DFT - e_GW = f(e_GW)
+        const double c = sx_vxc[i_qp] + _dft_energies[_opt.qpmin + i_qp];
+        Eigen::VectorXd xx_cur = xx.col(i_qp);              // lhs
+        Eigen::VectorXd fx_cur = fx.col(i_qp).array() + c;  // rhs
+        Eigen::VectorXd gx_cur = fx_cur - xx_cur;           // target
+        // Find best root
+        double root_value_max = 0.0;
+        double root_score_max = -1.0;
+        int root_idx = 0;
+        for (Index i_x = 0; i_x < nx - 1; ++i_x) {  // TODO: Loop only over sign-changes
+          if (gx_cur[i_x] * gx_cur[i_x + 1] < 0.0) {  // We have a sign change
+            // Estimate the root
+            double root_value_cur;
+            if (root_value_method == 0) {  // Average
+              root_value_cur = (xx_cur[i_x] + xx_cur[i_x + 1]) / 2.0;
+            } else if (root_value_method == 1) {  // Fixed-point
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Condition 'root_value_method==1' is always true](https://app.codacy.com/app/votca-package/xtp/pullRequest?prid=4286871)